incompatible type error on CoreSchema by mypy

PR #528 added a workaround for mypy's performance problem.
But the workaround code causes the incompatible-type error when we assign a concrete schema to a `CoreSchema` variable or argument.
```python
schema: core_schema.CoreSchema = core_schema.str_schema()
```

This is particularly problematic in defining a custom type.

**code**
```python
from typing import Any

from pydantic_core import core_schema


class MyValidationType:
    @classmethod
    def __get_pydantic_core_schema__(
        cls,
        **_kwargs: Any,
    ) -> core_schema.AfterValidatorFunctionSchema:
        return core_schema.field_after_validator_function(
            function=lambda _s, _info: MyValidationType(),  # some validation
            schema=core_schema.str_schema(),  # emits incompatible-type error
        )
```

**result**
```
% pipenv install mypy "pydantic>=2.0a1"
% pipenv run mypy test.py
test.py:14: error: Argument "schema" to "field_after_validator_function" has incompatible type "StringSchema"; expected "Dict[str, Any]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**environment**
```
% pipenv run pip list
Package           Version
----------------- -------
annotated-types   0.4.0
mypy              1.2.0
mypy-extensions   1.0.0
pip               23.0.1
pydantic          2.0a1
pydantic_core     0.23.1
setuptools        67.5.1
tomli             2.0.1
typing_extensions 4.5.0
wheel             0.38.4

% pipenv run python --version
Python 3.10.6
```

--- a/pydantic_core/core_schema.py
+++ b/pydantic_core/core_schema.py
@@ -1,6 +1,7 @@
 from __future__ import annotations as _annotations
 
 import sys
+from collections.abc import Mapping
 from datetime import date, datetime, time, timedelta
 from typing import Any, Callable, Dict, List, Optional, Set, Type, Union
 
@@ -3472,7 +3473,7 @@ if not MYPY:
         DefinitionReferenceSchema,
     ]
 elif False:
-    CoreSchema: TypeAlias = Dict[str, Any]
+    CoreSchema: TypeAlias = Mapping[str, Any]
 
 
 # to update this, call `pytest -k test_core_schema_type_literal` and copy the output

